p42/src/song.rs

```diff
@@ -0,0 +1,175 @@
+use core::panic;
+use std::io::{Read, Write};
+
+static DAYS: [&str; 12] = [
+    "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth",
+    "eleventh", "twelfth",
+];
+
+static GIFTS: [&str; 12] = [
+    " partridge in a pear tree",
+    "Two turtle doves,",
+    "Three French hens,",
+    "Four calling birds,",
+    "Five golden rings,",
+    "Six geese a-laying,",
+    "Seven swans a-swimming,",
+    "Eight maids a-milking,",
+    "Nine ladies dancing,",
+    "Ten lords a-leaping,",
+    "Eleven pipers piping,",
+    "Twelve drummers drumming,",
+];
+
+static LEFT: &str = "On the ";
+static RIGHT: &str = " day of Christmas,";
+static SECOND: &str = "my true love gave to me";
+static V1: &str = "A";
+static V2: &str = "And a";
+
+#[derive(Debug, Clone, Default, Copy)]
+pub struct SongIter {
+    pub day: usize,
+    pub line: usize,
+}
+
+impl Iterator for SongIter {
+    type Item = String;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.day == 12 {
+            return None;
+        }
+        let last = self.day + 2;
+        let result = match self.line {
+            0 => {
+                format!("{}{}{}\n", LEFT, DAYS[self.day], RIGHT)
+            }
+            1 => {
+                format!("{}\n", SECOND)
+            }
+            l if l < last => {
+                format!("{}\n", GIFTS[last - l])
+            }
+            l if l == last => match self.day {
+                0 => {
+                    format!("{}{}{}\n", V1, GIFTS[0], ".")
+                }
+                d if d < 12 => {
+                    format!("{}{}{}\n", V2, GIFTS[0], ".")
+                }
+                12 => {
+                    format!("{}{}{}\n", V2, GIFTS[0], "!")
+                }
+                _ => {
+                    panic!("This should never happen");
+                }
+            },
+            l if l == last + 1 => '\n'.to_string(),
+            _ => {
+                panic!("This should never happen");
+            }
+        };
+        self.line += 1;
+        if self.line > last + 1 {
+            self.day += 1;
+            self.line = 0;
+        }
+        Some(result)
+    }
+}
+
+pub fn numbered_lines() -> impl Iterator<Item = String> {
+    let song_iter = SongIter::default();
+    song_iter
+        .enumerate()
+        .map(|(i, line)| format!("{}: {}", i + 1, line))
+}
+
+// generic iterator wrapper to duplicate values N times
+pub struct Repeat<T: Iterator<Item: Clone>> {
+    pub iter: T,
+    pub n: usize,
```

Comment:
These fields can be private. Also note that `n` should be a const generic parameter.